File: python/src/d_micro_api.py
"""
Use pytest to test each case (even exceptions)
You should use FastAPI test client:
https://fastapi.tiangolo.com/tutorial/testing/

You will need to mock function is_admin_user
There are different ways to do this:
https://fastapi.tiangolo.com/advanced/testing-dependencies/
https://pytest-mock.readthedocs.io/en/latest/usage.html
"""
from typing import Annotated

from fastapi import FastAPI, HTTPException, Security, status
from pydantic import BaseModel

from lib.is_admin_user import is_admin_user

app = FastAPI()


class Message(BaseModel):
    message: str


@app.get("/hello/{name}")
async def hello(name: str) -> Message:
    if name.capitalize() != name:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="A name must starts with a capital letter",
        )

    return Message(message=f"Hello {name}!")


@app.get("/admin/hello")
async def hello_auth(is_admin: Annotated[bool, Security(is_admin_user)]) -> Message:
    if not is_admin:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN, detail="User must be admin"
        )

    return Message(message="Hello Admin!")

File: javascript/src/d-micro-api.ts
/**
 * Use jest to test each case (even exceptions)
 * You should use Fastify http injection:
 * https://fastify.dev/docs/v1.14.x/Documentation/Testing/
 *
 * You will need to mock function isAdminUser
 * https://jestjs.io/docs/jest-object#jestspyonobject-methodname
 * https://jestjs.io/docs/mock-function-api#mockfnmockresolvedvaluevalue
 */

import Fastify, {
  FastifyInstance,
  FastifyReply,
  FastifyRequest
} from 'fastify';

import { isAdminUser } from '../lib/is-admin-user';

function capitalize(word: string) {
  return word.charAt(0).toUpperCase() + word.slice(1);
}

export interface Message {
  message: string;
}

export function buildApp(): FastifyInstance {
  const fastify = Fastify();

  fastify.get(
    '/hello/:name',
    function (
      request: FastifyRequest<{ Params: { name: string } }>,
      reply: FastifyReply
    ) {
      const { name } = request.params;
      if (capitalize(name) != name) {
        reply
          .code(400)
          .send({ detail: 'A name must starts with a capital letter' });
      }

      reply.send({ message: `Hello ${name}!` });
    }
  );

  fastify.get('/admin/hello', async function (_, reply: FastifyReply) {
    const isAdmin = await isAdminUser();
    if (!isAdmin) {
      reply.code(403).send({ detail: 'User must be admin' });
    }

    reply.send({ message: 'Hello Admin!' });
  });

  return fastify;
}
